IABTechLab/uid2-self-serve-portal
# Improve audit trail pattern and increase coverage

## Overview
- Establish new pattern which reduces boilerplate to write audit logs
- Use this new pattern for all existing audit logs and also for features that lacked any audit logs (i.e. team member management)
- Simplify audit trail entity and service to be more generic
- Re-add participantId as a column in the audit trail table. Migrate existing participantIds from the eventData column into the participantId column.
- Some refactoring of the userController into the userService
- Fix a bug which meant that team members could not be updated

## The new audit trail pattern
- See `AuditTrail.ts` for the simplified entity, with less type-safety for the `eventData` field.
- See `auditTrailService.ts`, where many functions have been generalised into a single function. It also handles errors and updates the `succeeded` flag.
- There's many usages of the new pattern across the router and service files. Most of these changes live in `participantsRouter.ts`, and there are pretty much the same. Look at the code for deleting an API key as a fairly typical example of the new pattern.
- Any new features will just require adding to the `AuditTrailEvents` enum, then calling the existing function in `auditTrailService.ts`. No longer requires a new type, audit function as well as the need to manually update the succeeded flag. Just stick the event-specific data straight into the `eventData` field and move on.

## Manual testing
1. Run the latest db migration
1. Perform any action in the portal (bar Email Contact management). 
  a. Test Team Member management, particularly editing a team member
1. Observe the audit trails table and see that the information is accurate. Particularly userId, userEmail, event, eventData,  succeeded and participantId.
